app/controllers/msa_components_controller.rb

```diff
@@ -28,16 +34,19 @@ def create
   end
 
   def update
-    msa_component = MsaComponent.find_by_id(params[:id])
-    msa_component.team_id = params.dig(:component, :team_id)
-    event = ChangeComponentEvent.create(
-      component: msa_component,
-    )
-    unless event.valid?
-      error_message = event.errors.full_messages
-      flash[:notice] = error_message
+    @component = MsaComponent.find_by_id(params[:id])
+    @component.assign_attributes(component_params)
+    @event = ChangeComponentEvent.create(component: @component)
+
+    if @event.valid?
+      redirect_to msa_components_path
+    else
+      Rails.logger.info(@event.errors.full_messages)
+      @teams = Team.all
+      @hub_environments = Rails.configuration.hub_environments.keys
+
+      render :edit
```

Comment:
There are some helper methods that would remove the need to double up the logic in edit and update.

```
include ControllerConcern
include ComponentConcern

def update
  component_type = component_name_from_params(params)
  @component = klass_component(component_type)
...
end
```

Removes the need to have component specific methods.

Reply:
Unless I am misunderstanding, I am not sure this would be better than writing `MsaComponent.find_by_id(params[:id])` as we have the separate `sp` and `msa` controllers to work with the rails naming convention for views and routing etc. 

Maybe (for a separate task) it could be possible to refactor into one controller... i'm not sure how it would work with the routes, but I don't know if it would be worth it in the sense that its probably better to make things a bit more readable and to make it easier for people to navigate around the project.